### PR TITLE
gitlab ci install binary deps faster

### PR DESCRIPTION
**Pull request by haampie**
This PR does a few things:

- Parallel install of binary dependencies from buildcache (with redundant build edges pruned in the DAG)
- Improve legibility:
  1. Reduce verbosity installing buildcache deps, so that users get to see the build log for the source build
  2. Show timers for binary installation, so we get more insight into why a job is slow.
  3. Use GNU Make's output sync feature to get orderly output when installing deps in parallel
  4. Remove a few `spack -d` flags, so we get to see the build logs before hitting the max log size.

Screenshot from 2022-10-20 11-42-10
...
Screenshot from 2022-10-20 11-44-14

This PR is built on top of these parts:

- #33305
- #33315 
- #33361 
- #33363
- #33367

### CODE REVIEW

**Comment by haampie** [comment]
For reference:

- https://gitlab.spack.io/spack/spack/-/jobs/3618684 (Duration: 22 minutes 17 seconds, before this PR)
- https://gitlab.spack.io/spack/spack/-/jobs/3624534 (Duration: 8 minutes 11 seconds, after this PR)

or

- https://gitlab.spack.io/spack/spack/-/jobs/3618685 (Duration: 27 minutes 53 seconds, before this PR)
- https://gitlab.spack.io/spack/spack/-/jobs/3624526 (Duration: 8 minutes 53 seconds, after this PR)

**Comment by haampie** [comment]
@trws funnily enough we're hitting a broken `gmake -Orecurse` in Ubuntu 20.04 based containers :D I have no clue what they're doing... maybe they have some patches that broke it or so. The flag effectively has no effect. Fixed in a later version of Ubuntu.

**Comment by haampie** [comment]
Unfortunately the GNU make situation in our builders is rather sad:

- macOS system make is 3.81(too old for --output-sync)
- Amazon Linux 2 has gmake 3.82 (too old for --output-sync)
- Ubuntu 20.04 has a broken --output-sync gmake

Bootstrapping gmake is an option...

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by trws** [comment]
> Unfortunately the GNU make situation in our builders is rather sad:
> 
>     * macOS system make is 3.81(too old for --output-sync)
> 
>     * Amazon Linux 2 has gmake 3.82 (too old for --output-sync)
> 
>     * Ubuntu 20.04 has a broken --output-sync gmake
> 
> 
> Bootstrapping gmake is an option...

lets do that, or grab a statically linked one, or upgrade it from the package managers, or something.

**Comment by haampie** [comment]
@spackbot run pipeline

(@zackgalbreath @scottwittenburg: seems like the above has triggered on the old HEAD of this PR -- I've force pushed and if you follow the link below, you get Merge 253ae0b0 into 0de1f989 but the force pushed commit was 0e8b9fa as you can see above).

**Comment by haampie** [comment]
It's a one-liner to install julia's cross-compiled version, more people should use julia as a package manager

**Comment by haampie** [comment]
@spackbot fix style

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@scottwittenburg / @zackgalbreath it's apparently impossible to get a pipeline to run on the latest commit if it's in "Pending — waiting for base develop commit pipeline to succeed" mode? It continues to pick a commit from before a force push, even after pushing an ordinary commit in the meantime :/

**Comment by haampie** [comment]
@scottwittenburg / @kwryankrattiger can you review? Notice that a more recent GNU Make is not a hard requirement, it's just that it improves legibility as parallel output appears nice and orderly. The stop-gap solution of using Julia's pre-built binaries can be improved by moving it into the relevant docker images, but I think it's OK to have this solution for now.

https://gitlab.spack.io/spack/spack/-/pipelines/195792 <- this was the last pipeline before I removed the commit 395d95f1d134b69b7035c3859be633e97daab655 that was meant to trigger Gitlab CI.

**Comment by zackgalbreath** [comment]
> @scottwittenburg / @zackgalbreath it's apparently impossible to get a pipeline to run on the latest commit if it's in "Pending — waiting for base develop commit pipeline to succeed" mode? It continues to pick a commit from before a force push, even after pushing an ordinary commit in the meantime :/

FYI we just updated our sync script to remove this bit of functionality. Going forward, it should no longer block PRs with that "waiting for base develop commit pipeline to succeed" status.

File: lib/spack/spack/cmd/ci.py
Comment (scottwittenburg) [inline lib/spack/spack/cmd/ci.py:545]: Why did I think the runner images had everything and didn't need any bootstrapping? I suppose spack moves so fast that's not a practical goal?

**Comment by scottwittenburg** [comment]
This seems like it could be a great improvement, at least until more improvements to the underlying problems can land in spack (e.g. parallel/faster fetching of binaries, faster binary relocation, etc).

One thing that occurred to me is that we make these specific cpu requests to assist kubernetes in scheduling rebuild jobs on pods, we try to put them in every mapping entry, e.g. [here](https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml#L434), and make them match what we've seen them to require historically.     

I see this PR is basing make parallelism on `$nproc`, but these requests aren't limits, so I wonder if we need to worry about oversubscribing the instances we get run on?  

@zackgalbreath

File: lib/spack/spack/cmd/ci.py
Comment (haampie) [reply lib/spack/spack/cmd/ci.py:545]: It's part of https://github.com/spack/spack/pull/32262, some distros provide patchelf versions that are a decade old, and manipulating binaries is very bug-prone, so it makes sense to stick to a recent patchelf.

**Comment by haampie** [comment]
As far as I know nproc is aware of cgroups, for example:

```
taskset -c 0-1 nproc # nproc is restricted to cores 0 and 1
2
taskset -c 0-1 nproc --all
16
```
but idk how things are with kubernetes, since they have these fractional cpus etc

**Comment by scottwittenburg** [comment]
> but idk how things are with kubernetes, since they have these fractional cpus etc

IIUC the idea with Kubernetes is to tell it what kind of resources each job needs, and let it figure out how to pack jobs onto instances.  We can also give it resource limits, and if we do that for cpu, I think it throttles the processes to make sure they don't go over the limits we specify.  I'm not sure we want that though, maybe with cpu we just don't need to worry as much as with memory (where the risk is job death by OOM).

Opinions @zackgalbreath?

**Comment by haampie** [comment]
If you haven't configured `config:build_jobs` the same oversubscription issue would be true for `spack install` itself, since Spack determines the number of jobs with `os.sched_getaffinity(0)` which is exactly what `nproc` does.

```
taskset -c 0-1 spack python -c 'spack.util.cpus.cpus_available()'
2
```

That said, Spack's relocation logic may not be using composable threading (I believe it fires up a potentially large thread pool?). I guess we'll find out if this is an issue by trying... -- would be nice if spack was jobserver aware, i made an issue about that: #33014

**Comment by trws** [comment]
It's also aware of weak binding without cgroups, which is what taskset uses.  As long as only one spack install is allowed to actually *build* something at a time, this wont change the risks.  If there's a top-level `make -j 5` that runs multiple concurrent installs that run builds, rather than installing from the binary cache, then it's possible because of misbehaving build systems and linkers ignoring the jobserver, but I don't think we have that case here.

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@zackgalbreath / @scottwittenburg / @kwryankrattiger can you add a review?

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
Now seeing https://gitlab.spack.io/spack/spack/-/jobs/3862768 which is probably #31387, added #33478 and #33477 to improve at least the error... (I don't think it's a blocker since gitlab will retry and it happens at the start)

**Review by kwryankrattiger — DISMISSED** [review]
Two gripes that don't need to be fixed, mostly just notes for me later when we get to the CI module refactor
* Downloading the make isn't my favorite, but I don't have a better solution, so end of gripe 😞 
* Windows CI, whenever ever it happens, will likely not work well with this. So it will eventually need to be an option instead of always on. Other options would be adding Ninja (@haampie and I talked about this briefly) or adding finer-grained parallel install support in spack itself.

Thanks for your work on this @haampie

**Comment by haampie** [comment]
For windows we'd need to do some refactoring anyways, since we're now generating and executing a shell script. There's already room for `spack env depfile --generator ninja`, (currently it only supports `make` as a value). Downloading make is not great, but it can be moved in the base images.

**Comment by haampie** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
how do I make gitlab ci pass on this pr... Merge develop into here?

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
merging develop s.t. hopefully the openblas hash changes by in something more favorable...

**Comment by haampie** [comment]
@zackgalbreath what can I do to make this PR pass? Apparently merging develop doesn't change the openblas hashes. Why is it failing in this PR?

**Comment by haampie** [comment]
ping @scottwittenburg / @zackgalbreath / @kwryankrattiger, no clue how to proceed here.

**Comment by scottwittenburg** [comment]
I'm not sure what you can do while there is a broken `openblas` on `develop`.  You might get @tgamblin to merge this in spite of the failure if you think it's enough of an emergency to get this in.

**Comment by haampie** [comment]
It's not an emergency, but this is a ridiculous situation to be in... I don't need to fix openblas here, it's unrelated to my PR. Now my PR is blocked by it :|.

**Comment by scottwittenburg** [comment]
> but this is a ridiculous situation to be in

I agree it's unfortunate, sorry.  But what policy do you disagree with specifically?  That we fail PRs that generate broken hashes?  Or just that it affects this particular PR?  Keep in mind that one policy is not to prune "untouched" specs from a stack if the PR touched the `spack.yaml` for that stack, and this PR touches every `spack.yaml`.  So maybe you disagree with how we don't prune "untouched" specs when the affected stack's `spack.yaml` was touched?  

@tgamblin in case you're interested in discussing whether any of these policies should be revisited.

**Comment by haampie** [comment]
@zackgalbreath unmarked openblas cause it's a buildmachine related issue (apparently their makefile still checks the host arch even wtih DYNAMIC_ARCH=1, and the cpu type of the build node was something weird?)

**Comment by haampie** [comment]
@spackbot run pipeline

**Review by alalazo — APPROVED** [review]
We should eventually bootstrap `make` and `tar` too.

**Pull request merged**
